src/ports/postgres/modules/graph/sssp.py_in

```diff
@@ -0,0 +1,372 @@
+
+		plpy.execute(
+			""" CREATE TABLE {out_table} AS
+				SELECT {vertex_id}::INT AS {vertex_id},
+					CAST('Infinity' AS DOUBLE PRECISION) AS {weight},
+					CAST({INT_MAX} AS INT) AS parent
+				FROM {vertex_table} {distribution} """.format(**locals()))
+		plpy.execute(
+			""" CREATE TEMP TABLE {oldupdate}(
+				id INT, val DOUBLE PRECISION, parent INT)
+				{local_distribution}
+				""".format(**locals()))
+		plpy.execute(
+			""" CREATE TEMP TABLE {newupdate}(
+				id INT, val DOUBLE PRECISION, parent INT)
+				{local_distribution}
+				""".format(**locals()))
+		temp_table = unique_string(desp='temp')
+		sql = m4_ifdef(<!__HAWQ__!>,
+			""" CREATE TABLE {temp_table} (
+					{vertex_id} INT, {weight} DOUBLE PRECISION, parent INT)
+					{distribution};
+			""",  <!''!>)
+		plpy.execute(sql.format(**locals()))
+
+		sql_index = m4_ifdef(<!__POSTGRESQL__!>,
+			<!""" CREATE INDEX ON {out_table} ({vertex_id});
+				CREATE INDEX ON {oldupdate} (id);
+				CREATE INDEX ON {newupdate} (id);
+			""".format(**locals())!>,
+			<!''!>)
+		plpy.execute(sql_index)
+
+		plpy.execute(
+			""" INSERT INTO {oldupdate}
+				VALUES({source_vertex},0,{source_vertex})
+			""".format(**locals()))
+
+		v_cnt = plpy.execute("SELECT count(*) FROM {vertex_table}".
+			format(**locals()))[0]['count']
+		for i in range(0,v_cnt+1):
+
+			sql = m4_ifdef(<!__HAWQ__!>,
+				<!"""
+				TRUNCATE TABLE {temp_table};
+				INSERT INTO {temp_table}
+					SELECT *
+					FROM {out_table}
+					WHERE {out_table}.{vertex_id} NOT IN (
+						SELECT {oldupdate}.id FROM {oldupdate})
+					UNION
+					SELECT * FROM {oldupdate};
+				DROP TABLE {out_table};
+				ALTER TABLE {temp_table} RENAME TO {out_table};
+				CREATE TABLE {temp_table} (
+					{vertex_id} INT, {weight} INT, parent INT) {distribution};
+				"""!>,
+				<!"""
+				UPDATE {out_table} SET
+				{weight}=oldupdate.val,
+				parent=oldupdate.parent
+				FROM
+				{oldupdate} AS oldupdate
+				WHERE
+				{out_table}.{vertex_id}=oldupdate.id
+				"""!>)
+			plpy.execute(sql.format(**locals()))
+
+			plpy.execute("TRUNCATE TABLE {0}".format(newupdate))
+			sql = ("""
+				INSERT INTO {newupdate}
+				SELECT DISTINCT ON (message.id) message.id AS id,
+					message.val AS val,
+					message.parent AS parent
+				FROM  {out_table} AS out_table,
+					(SELECT edge_table.{dest} AS id, x.val AS val,
+						oldupdate.id AS parent
+					FROM {oldupdate} AS oldupdate, {edge_table} AS edge_table,
+						(SELECT edge_table.{dest} AS id,
+							min(oldupdate.val + edge_table.{weight}) AS val
+						FROM {oldupdate} AS oldupdate,
+							{edge_table} AS edge_table
+						WHERE edge_table.{src}=oldupdate.id
+						GROUP BY edge_table.{dest}) x
+					WHERE edge_table.{src} = oldupdate.id
+						AND edge_table.{dest} = x.id
+						AND ABS(oldupdate.val + edge_table.{weight} - x.val) <
+							{EPSILON}
+					) AS message
+				WHERE message.id = out_table.{vertex_id}
+					AND message.val<out_table.{weight}
+					""".format(**locals()))
```

Comment:
I don't think this WHERE and the DISTINCT are the right way to do this, but since I haven't been able to wrap my head around the intent here I don't know.